lib/using_restricted_array.rb

```diff
@@ -6,56 +6,145 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def length(array)
-  raise NotImplementedError
+  n = 0
+  until array[n] == nil
+    n += 1
+  end
+  return n
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  if array[0] == nil 
+    puts "[]"
+  else
+    puts "["
+    n = 0
+    m = length(array)
+    while n < m
+      puts "#{array[n]}, "
+      n += 1
+    end
+    puts "#{array[-1]}"
+    puts "]"
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  n = 0
+  found = 0
+  while n < length
+    if value_to_find == array[n]
+      return true
+      found += 1
+    end
+    n += 1
+  end
+
+  if found == 0
```

Comment:
You can simply return false if this line is encountered, there's no need for an if statement.